worker/transfer.go

```diff
@@ -125,46 +128,107 @@ func parallelUploadSlab(ctx context.Context, sp storeProvider, shards [][]byte,
 		span.End()
 	}
 
-	// spawn workers and send initial requests
-	hostIndex := 0
+	// helper to track used hosts.
+	hostUsed := make([]bool, len(contracts))
+	nextHost := func() int {
+		for i := range hostUsed {
+			if !hostUsed[i] {
+				hostUsed[i] = true
+				return i
+			}
+		}
+		return -1
+	}
+
+	// helper to determine whether a request is finished.
+	isReqFinished := func(r req) bool {
+		select {
+		case <-r.finishedCtx.Done():
+			return true
+		default:
+			return false
+		}
+	}
+
+	// helper function for launching worker given an existing request. Returns
+	// true if a worker was launched for the request or if the request was
+	// already finished.
 	inflight := 0
-	for i := range shards {
-		go worker(req{contracts[hostIndex], i})
-		hostIndex++
+	launchWorker := func(r req) bool {
+		if isReqFinished(r) {
+			return true
+		}
+		hostIndex := nextHost()
+		if hostIndex == -1 {
+			return false
+		}
+		go worker(req{r.finishedCtx, r.finishedFn, hostIndex, r.shardIndex})
 		inflight++
+		return true
+	}
+
+	// spawn workers and send initial requests
+	for i := range shards {
+		finishedCtx, finishedFn := context.WithCancel(ctx)
+		if !launchWorker(req{finishedCtx, finishedFn, -1, i}) {
+			panic("failed to launch worker for initial shard - should never happen")
+		}
 	}
 
 	// collect responses
 	var errs HostErrorSet
 	sectors := make([]object.Sector, len(shards))
 	rem := len(shards)
+	var neededOverdrive []req
 	for rem > 0 && inflight > 0 {
 		resp := <-respChan
 		if !errors.Is(resp.err, errUploadSectorTimeout) {
 			inflight--
 		}
 
-		if resp.err != nil {
-			errs = append(errs, &HostError{resp.req.contract.HostKey, resp.err})
-			// try next host
-			if hostIndex < len(contracts) {
-				go worker(req{contracts[hostIndex], resp.req.shardIndex})
-				hostIndex++
-				inflight++
+		// Decide whether to reuse the host or not. We only do that if the
+		// worker finished executing and failed due to the request being
+		// finished already.
+		if resp.err != nil && !errors.Is(resp.err, errUploadSectorTimeout) {
+			// Otherwise remember the error and reuse the host if the request
+			// was already finished.
+			errs = append(errs, &HostError{contracts[resp.req.hostIndex].HostKey, resp.err})
+			if isReqFinished(resp.req) {
+				hostUsed[resp.req.hostIndex] = false
 			}
+		}
+
+		if errors.Is(resp.err, errUploadSectorTimeout) {
+			// for each slow host we eventually launch an overdrive worker
```

Comment:
That's why it says "eventually". Because the cap only affects currently active overdrive workers. So as overdrive workers finish, we may be able to launch new ones.

Reply:
I was `nit`ting about the edge case where the max overdrive is set to zero 😛 . But leave it it really doesn't matter